src/main/java/com/epam/aidial/core/controller/RateResponseController.java

```diff
@@ -0,0 +1,178 @@
+        log.info("Connected to origin: {}", proxyRequest.connection().remoteAddress());
+
+        HttpServerRequest request = context.getRequest();
+        context.setProxyRequest(proxyRequest);
+
+        ProxyUtil.copyHeaders(request.headers(), proxyRequest.headers());
+
+        Buffer proxyRequestBody = context.getRequestBody();
+        proxyRequest.putHeader(HttpHeaders.CONTENT_LENGTH, Integer.toString(proxyRequestBody.length()));
+
+        proxyRequest.send(proxyRequestBody)
+                .onSuccess(this::handleProxyResponse)
+                .onFailure(this::handleProxyRequestError);
+    }
+
+    /**
+     * Called when proxy received the response headers from the origin.
+     */
+    private void handleProxyResponse(HttpClientResponse proxyResponse) {
+        log.info("Received response header from origin: status={}, headers={}", proxyResponse.statusCode(),
+                proxyResponse.headers().size());
+
+        if (proxyResponse.statusCode() == HttpStatus.TOO_MANY_REQUESTS.getCode()) {
+            sendRequest(); // try next
+            return;
+        }
+
+        BufferingReadStream proxyResponseStream = new BufferingReadStream(proxyResponse,
+                ProxyUtil.contentLength(proxyResponse, 1024));
+
+        context.setProxyResponse(proxyResponse);
+        context.setResponseStream(proxyResponseStream);
+
+        HttpServerResponse response = context.getResponse();
+        response.setChunked(true);
+        response.setStatusCode(proxyResponse.statusCode());
+        ProxyUtil.copyHeaders(proxyResponse.headers(), response.headers());
+
+        proxyResponseStream.pipe()
+                .endOnFailure(false)
+                .to(response)
+                .onSuccess(ignored -> handleResponse())
+                .onFailure(this::handleResponseError);
+    }
+
+    /**
+     * Called when proxy sent response from the origin to the client.
+     */
+    private void handleResponse() {
+        Buffer proxyResponseBody = context.getResponseStream().getContent();
+        context.setResponseBody(proxyResponseBody);
+        proxy.getLogStore().save(context);
+    }
+
+    /**
+     * Called when proxy failed to receive request body from the client.
+     */
+    private void handleRequestBodyError(Throwable error) {
+        log.warn("Failed to receive client body: {}", error.getMessage());
+        context.respond(HttpStatus.UNPROCESSABLE_ENTITY, "Failed to receive body");
+    }
+
+    /**
+     * Called when proxy failed to connect to the origin.
+     */
+    private void handleProxyConnectionError(Throwable error) {
+        log.warn("Can't connect to origin: {}", error.getMessage());
+        sendRequest(); // try next
+    }
+
+    /**
+     * Called when proxy failed to send request to the origin.
+     */
+    private void handleProxyRequestError(Throwable error) {
+        log.warn("Can't send request to origin: {}", error.getMessage());
+        sendRequest(); // try next
```

Comment:
Do not retry. Otherwise you get into an infinite loop of getting this error and retrying again and again.

Reply:
fixed.